src/linker/Linker.Steps/SweepStep.cs

```diff
@@ -728,16 +550,348 @@ protected virtual void ElementRemoved (IMetadataTokenProvider element)
+
+				if (provider.MarshalInfo is CustomMarshalInfo cmi)
+					UpdateScopeOfTypeReference (cmi.ManagedType);
+			}
+
+			void UpdateCustomAttributesTypesScopes (ICustomAttributeProvider customAttributeProvider)
+			{
+				if (!customAttributeProvider.HasCustomAttributes)
+					return;
+
+				foreach (var ca in customAttributeProvider.CustomAttributes)
+					UpdateForwardedTypesScope (ca);
+			}
+
+			void UpdateForwardedTypesScope (CustomAttribute attribute)
+			{
+				UpdateMethodReference (attribute.Constructor);
+
+				if (attribute.HasConstructorArguments) {
+					foreach (var ca in attribute.ConstructorArguments)
+						UpdateForwardedTypesScope (ca);
+				}
+
+				if (attribute.HasFields) {
+					foreach (var field in attribute.Fields)
+						UpdateForwardedTypesScope (field.Argument);
+				}
+
+				if (attribute.HasProperties) {
+					foreach (var property in attribute.Properties)
+						UpdateForwardedTypesScope (property.Argument);
+				}
+			}
+
+			void UpdateForwardedTypesScope (CustomAttributeArgument attributeArgument)
+			{
+				UpdateScopeOfTypeReference (attributeArgument.Type);
+
+				switch (attributeArgument.Value) {
+				case TypeReference tr:
+					UpdateScopeOfTypeReference (tr);
+					break;
+				case CustomAttributeArgument caa:
+					UpdateForwardedTypesScope (caa);
+					break;
+				case CustomAttributeArgument[] array:
+					foreach (var item in array)
+						UpdateForwardedTypesScope (item);
+					break;
+				}
+			}
+
+			void UpdateScopeOfTypeReference (TypeReference type)
+			{
+				if (type == null)
+					return;
+
+				if (updated.Contains (type))
+					return;
+
+				updated.Add (type);
+
+				// Can't update the scope of windows runtime projections
+				if (type.IsWindowsRuntimeProjection)
+					return;
+
+				switch (type) {
+				case GenericInstanceType git:
+					UpdateScopeOfTypeReference (git.ElementType);
+					foreach (var ga in git.GenericArguments)
+						UpdateScopeOfTypeReference (ga);
+					return;
+				case FunctionPointerType fpt:
+					UpdateScopeOfTypeReference (fpt.ReturnType);
+					if (fpt.HasParameters)
+						UpdateTypeScope (fpt.Parameters);
+					return;
+				case TypeSpecification ts:
+					UpdateScopeOfTypeReference (ts.ElementType);
+					return;
+#if FEATURE_ILLINK
+				case TypeDefinition:
+				case GenericParameter:
+#else
+				case TypeDefinition _:
+				case GenericParameter _:
+#endif
+					// Nothing to update
+					return;
+				}
+
+				//
+				// Resolve to type definition to remove any type forwarding imports
+				//
+				TypeDefinition td = type.Resolve ();
+				if (td == null) {
+					//
+					// This can happen when not all assembly refences were provided and we
+					// run in `--skip-unresolved` mode. We cannot fully sweep and keep the
+					// original assembly reference
+					//
+					var anr = (AssemblyNameReference) type.Scope;
+					type.Scope = importer.ImportReference (anr);
+					return;
+				}
+
+				var tr = assembly.MainModule.ImportReference (td);
```

Comment:
For a typeref that resolves to an unused assembly, I think this will keep the assembly reference. We [used to](https://github.com/mono/linker/blob/master/src/linker/Linker.Steps/SweepStep.cs#L295) remove such unused assemblyrefs (at the expense of leaving inconsistent typerefs in unused code). Is this expected?

Reply:
The logic here is different, the processing is running only on types/members which are marked (anything unused was removed from the collection before this logic was entered)

Reply:
Couldn't we get here for unmarked members in a `copyused` assembly through `SweepAssemblyReferences`?

Reply:
We could and in that case we don't care about marked/unmarked because for copy/copyused we want to keep all members, essentially this code processed assembly in its final form.